File: app/src/main/java/org/ngbp/jsonrpc4jtestharness/core/ws/MiddlewareWebSocketClient.java
package org.ngbp.jsonrpc4jtestharness.core.ws;

import org.eclipse.jetty.client.HttpClient;
import org.eclipse.jetty.util.ssl.SslContextFactory;
import org.eclipse.jetty.websocket.api.Session;
import org.eclipse.jetty.websocket.client.WebSocketClient;

import java.net.URI;
import java.util.concurrent.Future;

public class MiddlewareWebSocketClient {

    public void start() {
        URI uri = URI.create("wss://localhost:9999/atscCmd");

        SslContextFactory sslContextFactory = new SslContextFactory.Client();
        sslContextFactory.setTrustAll(true);
        HttpClient http = new HttpClient(sslContextFactory);

        WebSocketClient client = new WebSocketClient(http);
        try {
            try {
                client.start();
                // The socket that receives events
                MiddlewareWebSocket socket = new MiddlewareWebSocket();
                // Attempt Connect
                Future<Session> fut = client.connect(socket,uri);
                // Wait for Connect
                Session session = fut.get();
                // Send a message
                session.getRemote().sendString("Hello");
                // Close session
                session.close();
            } finally {
                client.stop();
            }
        }
        catch (Throwable t) {
            t.printStackTrace(System.err);
        }
    }
}

File: app/src/main/java/org/ngbp/jsonrpc4jtestharness/core/ws/MiddlewareWebSocketClient.kt
package org.ngbp.jsonrpc4jtestharness.core.ws

import org.eclipse.jetty.client.HttpClient
import org.eclipse.jetty.util.ssl.SslContextFactory
import org.eclipse.jetty.websocket.client.WebSocketClient
import org.ngbp.jsonrpc4jtestharness.rpc.processor.IRPCProcessor
import java.net.URI

class MiddlewareWebSocketClient(private var rpcProcessor: IRPCProcessor) {
    fun start() {
        val uri = URI.create("ws://localhost:9998/echo")
        val sslContextFactory: SslContextFactory = SslContextFactory.Client()
        sslContextFactory.isTrustAll = true
        val http = HttpClient(sslContextFactory)
        val client = WebSocketClient(http)
        try {
            try {
                client.start()
                // The socket that receives events
                val socket = MiddlewareWebSocket(rpcProcessor, null)
                // Attempt Connect
                val fut = client.connect(socket, uri)
                // Wait for Connect
                val session = fut.get()
                // Send a message
                session.remote.sendString("{\"jsonrpc\":\"2.0\",\"method\":\"org.atsc.subscribe\",\"params\":{\"msgType\":[\"All\"]},\"id\":1}")
                // Close session
                session.close()
            } finally {
                client.stop()
            }
        } catch (t: Throwable) {
            t.printStackTrace(System.err)
        }
    }
}